### ISSUE TITLE
Result node is incorrectly "zapped"

### ISSUE THREAD

**Issue by hlinnaka** [greenplum-db/gpdb#7465]
On 5X_STABLE:

```
CREATE TABLE zapfail
(
  partkey TEXT,
  filler TEXT,
  col_len INTEGER
)
PARTITION BY list (partkey)
(
  PARTITION foo VALUES ('foo'),
  PARTITION bar VALUES ('bar')
);

insert into zapfail(col_len, partkey) values ('3', 'foo');

select partkey from zapfail a where col_len = length(partkey);

ERROR:  invalid memory alloc request size 2138009450 (context 'ExecutorState') (tuptoaster.c:1690) (mcxt.c:473)
```

The error varies from one invocation to another. I've seen the above, and "relation not found" errors, as well as segfaults.

**Comment by hlinnaka** [comment]
This works in 6X_STABLE. I used git bisect to find the commit that fixed it: it was c613cabf34. Looking at the EXPLAIN VERBOSE output, it's easy to see that it's related to confusion on target lists, caused by "zapping" the Result node between the Gather and the Append:
```
postgres=# explain verbose select partkey from zapfail a where col_len = length(partkey);
                                    QUERY PLAN                                     
-----------------------------------------------------------------------------------
 Gather Motion 3:1  (slice1; segments: 3)  (cost=0.00..1202.00 rows=67 width=36)
   Output: a.partkey
   ->  Append  (cost=0.00..1202.00 rows=23 width=36)
         ->  Seq Scan on zapfail_1_prt_foo a  (cost=0.00..601.00 rows=12 width=36)
               Output: a.partkey, a.col_len
               Filter: col_len = length(partkey)
         ->  Seq Scan on zapfail_1_prt_bar a  (cost=0.00..601.00 rows=12 width=36)
               Output: a.partkey, a.col_len
               Filter: col_len = length(partkey)
 Settings:  optimizer=off
 Optimizer status: legacy query optimizer
(11 rows)
```
Note that the targetlist of the Append node is "a.partkey, a.col_len" is different from the target list of the Gather node, "a.partkey" (the target list of the Append node is not shown explicitly, but it's the same as for the SeqScan nodes below). The Gather node gets confused about that. The sender side of the Motion sends the tuple with two attributes, but the receiving side tries to store it in a tuple slot with only one attribute.

To fix, I think we should backport commit c613cabf to 5X_STABLE. Alternatively, we could make the "zapping" code to be more strict, and skip the zapping if there's a Gather node above it. Or modify the Motion code to not get confused by that.  But backporting that seems simplest.

**Comment by hlinnaka** [comment]
Never mind, this was already fixed by https://github.com/greenplum-db/gpdb/pull/7367, commit 44d83e0b.
